src/epochs.lua

```diff
@@ -370,71 +359,62 @@ end
---- @return Epoch|nil # The created epoch, or nil if an epoch already exists for the index
-function epochs.createEpoch(timestamp, blockHeight, hashchain)
+--- @return PrescribedEpoch | nil # The created epoch, or nil if an epoch already exists for the index
+function epochs.createAndPrescribeNewEpoch(timestamp, blockHeight, hashchain)
 	assert(type(timestamp) == "number", "Timestamp must be a number")
 	assert(type(blockHeight) == "number", "Block height must be a number")
 	assert(type(hashchain) == "string", "Hashchain must be a string")
 
+	-- if before the epoch zero start timestamp, return nil
+	if timestamp < epochs.getSettings().epochZeroStartTimestamp then
+		print("Genesis epoch will start at: " .. epochs.getSettings().epochZeroStartTimestamp)
+		return nil
+	end
+
 	local epochIndex = epochs.getEpochIndexForTimestamp(timestamp)
 	if epochs.getEpoch(epochIndex) then
-		-- silently return
-		print("Epoch already exists for index: " .. epochIndex)
+		print("Epoch already exists for index, skipping creation: " .. epochIndex)
 		return nil -- do not return the existing epoch to prevent sending redundant epoch-created-notices
 	end
 
-	local prevEpochIndex = epochIndex - 1
-	local prevEpoch = epochs.getEpoch(prevEpochIndex)
-	-- if there is a previous epoch and it has not been distributed, we cannot create a new epoch. once the epoch has been distributed, it will be removed from the epoch registry
-	if prevEpoch and not prevEpoch.distributions.distributedTimestamp then
-		-- TODO: consider throwing an error here instead of silently returning
-		print(
-			"Distributions have not occurred for the previous epoch. A new epoch cannot be created until distribution for the previous epoch is complete: "
-				.. prevEpochIndex
-		)
-		return nil
-	end
-
-	local epochStartTimestamp, epochEndTimestamp, epochDistributionTimestamp =
-		epochs.getEpochTimestampsForIndex(epochIndex)
-	local prescribedObservers, weightedGateways = epochs.computePrescribedObserversForEpoch(epochIndex, hashchain)
-	local prescribedNames = epochs.computePrescribedNamesForEpoch(epochIndex, hashchain)
-	local activeGateways = gar.getActiveGatewaysBeforeTimestamp(epochStartTimestamp)
 	-- get the max rewards for each participant eligible for the epoch
+	local prescribedObservers, updatedGatewaysWithWeights =
+		epochs.computePrescribedObserversForEpoch(epochIndex, hashchain)
+	local prescribedNames = epochs.computePrescribedNamesForEpoch(epochIndex, hashchain)
 	local eligibleEpochRewards = epochs.computeTotalEligibleRewardsForEpoch(epochIndex, prescribedObservers)
-	-- snapshot the ARNS stats at the beginning of the epoch, does not account for any names that are created or expire during the epoch
+	local epochStartTimestamp, epochEndTimestamp = epochs.getEpochTimestampsForIndex(epochIndex)
 	local arnsStatsAtEpochStart = arns.getArNSStatsAtTimestamp(epochStartTimestamp)
+
+	-- update gateway weights for every new epoch, so its clear why the prescribed observers are selected
```

Comment:
Not sure what the assignment here has to do with "clarity" 🤔

Reply:
Does hoisting the weights assignment here before the epoch initialization below change any behavior?

Reply:
> Does hoisting the weights assignment here before the epoch initialization below change any behavior?
it does not
> Not sure what the assignment here has to do with "clarity" 🤔
can reword - but the freshly computed weights are saved in state, and viewable by operators for the epoch